pkg/controller/utils/capability.go

```diff
@@ -0,0 +1,286 @@
+		Controller:         pointer.BoolPtr(true),
+		BlockOwnerDeletion: pointer.BoolPtr(true),
+	}}
+	return def.CreateOrUpdateConfigMap(ctx, k8sClient, traitDefinition.Name, jsonSchema, ownerReference)
+}
+
+// CapabilityBaseDefinition is the base struct for CapabilityWorkloadDefinition and CapabilityTraitDefinition
+type CapabilityBaseDefinition struct {
+}
+
+// CreateOrUpdateConfigMap creates ConfigMap to store OpenAPI v3 schema or or updates data in ConfigMap
+func (def *CapabilityBaseDefinition) CreateOrUpdateConfigMap(ctx context.Context, k8sClient client.Client, definitionName string, jsonSchema []byte, ownerReferences []metav1.OwnerReference) error {
+	cmName := fmt.Sprintf("%s%s", types.CapabilityConfigMapNamePrefix, definitionName)
+	ns := types.DefaultKubeVelaNS
+	var cm v1.ConfigMap
+	var data = map[string]string{
+		types.OpenapiV3JSONSchema: string(jsonSchema),
+	}
+	err := k8sClient.Get(ctx, client.ObjectKey{Namespace: ns, Name: cmName}, &cm)
+	if err != nil && apierrors.IsNotFound(err) {
+		cm = v1.ConfigMap{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "ConfigMap",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            cmName,
+				Namespace:       ns,
+				OwnerReferences: ownerReferences,
+			},
+			Data: data,
+		}
+		err = k8sClient.Create(ctx, &cm)
+		if err != nil {
+			return fmt.Errorf(util.ErrUpdateCapabilityInConfigMap, definitionName, err)
+		}
+		return nil
+	} else if cm.Data[types.OpenapiV3JSONSchema] != string(jsonSchema) {
+		cm.Data = data
+		err := k8sClient.Update(ctx, &cm)
+		if err != nil {
+			return fmt.Errorf(util.ErrUpdateCapabilityInConfigMap, definitionName, err)
+		}
+		return nil
+	}
+	return nil
+}
+
+// getDefinition is the main function for GetDefinition API
+func getOpenAPISchema(capability types.Capability) ([]byte, error) {
+	openAPISchema, err := generateOpenAPISchemaFromCapabilityParameter(capability)
+	if err != nil {
+		return nil, err
+	}
+	swagger, err := openapi3.NewSwaggerLoader().LoadSwaggerFromData(openAPISchema)
+	if err != nil {
+		return nil, err
+	}
+	schemaRef := swagger.Components.Schemas["parameter"]
+	if schemaRef == nil {
+		return nil, fmt.Errorf(util.ErrGenerateOpenAPIV2JSONSchemaForCapability, capability.Name, nil)
+	}
+	schema := schemaRef.Value
+	fixOpenAPISchema("", schema)
+
+	parameter, err := schema.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	return parameter, nil
+}
+
+// generateOpenAPISchemaFromCapabilityParameter returns the parameter of a definition in cue.Value format
+func generateOpenAPISchemaFromCapabilityParameter(capability types.Capability) ([]byte, error) {
+	name := capability.Name
+	definitionCueName := fmt.Sprintf("%s.cue", name)
+	template, err := prepareParameterCue(name, capability.CueTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	// append context section in CUE string
+	template += mycue.BaseTemplate
+	dir, err := ioutil.TempDir("", "openapi")
+	//nolint:errcheck
+	defer os.RemoveAll(dir)
```

Comment:
do we really need to create a real file and delete it? Can we just use memory operation?

Reply:
The CUE API asked for a folder and a file to generate schema. I searched, but could not find an alternative way.

Reply:
```
var r cue.Runtime
cueInst, err := r.Compile("-", myCueTemplate)
...
GenOpenAPI(cueInst)
```